jsontableschema/validate.py

```diff
@@ -4,284 +4,110 @@
-                                  'is present with unclear application (field '
-                                  'is not an integer or a date)')
+                for k in instance['primaryKey']:
+                    field_names = [f['name'] for f in instance['fields']]
+                    if k not in field_names:
+                        yield exceptions.ValidationError(
+                            'A JSON Table Schema primaryKey value must be '
+                            'found in the schema field names'
+                        )
+
+        # the hash may contain a key `foreignKeys`
+        if isinstance(instance, dict) and instance.get('foreignKeys'):
+            for fk in instance['foreignKeys']:
+                # ensure that `foreignKey.fields` match field names
+                if isinstance(fk.get('fields'), compat.str):
+                    field_names = [f['name'] for f in instance['fields']]
+                    if fk.get('fields') not in field_names:
+                        yield exceptions.ValidationError(
+                            'A JSON Table Schema foreignKey.fields value must '
+                            'correspond with field names.'
+                        )
+                else:
+                    for field in fk.get('fields'):
+                        field_names = [f['name'] for f in instance['fields']]
+                        if field not in field_names:
+                            yield exceptions.ValidationError(
+                                'A JSON Table Schema foreignKey.fields value '
+                                'must correspond with field names.'
+                            )
+
+                # ensure that `foreignKey.reference.fields`
+                # matches outer `fields`
+                if isinstance(fk.get('fields'), compat.str):
+                    if not isinstance(fk['reference']['fields'], compat.str):
+                        yield exceptions.ValidationError(
+                            'A JSON Table Schema foreignKey.reference.fields '
+                            'must match field names.'
+                        )
+                else:
+                    if isinstance(fk['reference']['fields'], compat.str):
+                        yield exceptions.ValidationError(
+                            'A JSON Table Schema foreignKey.fields cannot '
+                            'be a string when foreignKey.reference.fields.'
+                            'is a string'
+                        )
+                    if not len(fk.get('fields')) == len(fk['reference']['fields']):
+                        yield exceptions.ValidationError(
+                            'A JSON Table Schema foreignKey.fields must '
+                            'contain the same number entries as '
+                            'foreignKey.reference.fields.'
+                        )
+
+
+validator = JSONTableSchemaValidator(json_table_schema)
```

Comment:
I'm not sure about this name, but this is what you'd end up using if you wanted to iterate over all errors by calling

    validator.iter_errors(schema)

but it's the same as the jsonschema interface.